File: frontend/src/types/sheet_agent.ts
export type SheetSourceType = 'csv' | 'xlsx' | 'google_sheets'

export type SheetDatasourceStatus = 'pending' | 'processing' | 'ready' | 'failed'

export type SheetDatasource = {
  id: string
  chat_id: string
  user_id: string
  source_type: SheetSourceType
  file_name: string | null
  storage_path: string | null
  file_size_bytes: number | null
  sheet_url: string | null
  sheet_id: string | null
  sheet_tab: string | null
  status: SheetDatasourceStatus
  row_count: number | null
  column_count: number | null
  column_names: string[] | null
  sheet_tabs: string[] | null
  error_message: string | null
  created_at: string
  updated_at: string
}

export type SheetUploadResponse = {
  datasource: SheetDatasource
}

export type ConnectGoogleSheetRequest = {
  chat_id: string
  sheet_url: string
  sheet_tab?: string | null
}

export type ConnectGoogleSheetResponse = {
  datasource: SheetDatasource
}

export type SheetQueryRequest = {
  datasource_id: string
  question: string
  chat_id?: string | null
  sheet_tab?: string | null
}

export type SheetQueryTableRow = {
  columns: string[]
  rows: string[][]
}

export type SheetQueryResponse = {
  chat_id: string
  user_message_id: string
  assistant_message_id: string
  datasource_id: string
  question: string
  answer: string
  table: SheetQueryTableRow | null
  execution_duration_ms: number
}

export type ListDatasourcesResponse = {
  items: SheetDatasource[]
}

File: backend/app/schemas/sheet_agent.py
import json
import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field, field_validator


SourceType = Literal["csv", "xlsx", "google_sheets"]
DatasourceStatus = Literal["pending", "processing", "ready", "failed"]


class SheetDatasourceRead(BaseModel):
    id: uuid.UUID
    chat_id: uuid.UUID
    user_id: uuid.UUID
    source_type: SourceType
    file_name: str | None
    storage_path: str | None
    file_size_bytes: int | None
    sheet_url: str | None
    sheet_id: str | None
    sheet_tab: str | None
    status: DatasourceStatus
    row_count: int | None
    column_count: int | None
    column_names: list[str] | None
    sheet_tabs: list[str] | None
    error_message: str | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}

    @classmethod
    def model_validate(cls, obj, **kwargs):
        # Deserialize JSON-encoded column/tab lists stored as Text
        if hasattr(obj, "__dict__"):
            data = {c.key: getattr(obj, c.key) for c in obj.__table__.columns}
            for field in ("column_names", "sheet_tabs"):
                raw = data.get(field)
                if isinstance(raw, str):
                    try:
                        data[field] = json.loads(raw)
                    except (ValueError, TypeError):
                        data[field] = None
            return cls(**data)
        return super().model_validate(obj, **kwargs)


class SheetUploadResponse(BaseModel):
    datasource: SheetDatasourceRead


class ConnectGoogleSheetRequest(BaseModel):
    chat_id: uuid.UUID
    sheet_url: str = Field(min_length=1, max_length=2000)
    sheet_tab: str | None = Field(default=None, max_length=255)

    @field_validator("sheet_url")
    @classmethod
    def validate_sheet_url(cls, v: str) -> str:
        if "docs.google.com/spreadsheets" not in v:
            raise ValueError("URL must be a valid Google Sheets URL")
        return v.strip()


class ConnectGoogleSheetResponse(BaseModel):
    datasource: SheetDatasourceRead


class SheetQueryRequest(BaseModel):
    datasource_id: uuid.UUID
    question: str = Field(min_length=1, max_length=4000)
    chat_id: uuid.UUID | None = None
    sheet_tab: str | None = None


class SheetQueryTableRow(BaseModel):
    columns: list[str]
    rows: list[list[str]]


class SheetQueryResponse(BaseModel):
    chat_id: uuid.UUID
    user_message_id: uuid.UUID
    assistant_message_id: uuid.UUID
    datasource_id: uuid.UUID
    question: str
    answer: str
    table: SheetQueryTableRow | None = None
    execution_duration_ms: int


class ListDatasourcesResponse(BaseModel):
    items: list[SheetDatasourceRead]
